### PR TITLE
[WIP] Major refactor of input/output specs and autonaming

### PR DESCRIPTION
**Pull request by oesteban**
### Make name_source available at BaseInterface level

This PR enables the functionality of name_source for any interface derived from BaseInterface. Previously, it only worked for CommandLine interfaces.

There's some previous discussion about this issue here #573. 
And we started working on this from here #1234.

I have moved the old ```_filename_from_source``` in ```CommandLine``` to ```BaseInterface```, with the new name of ```_resolve_namesource```.

The ```run()``` method and ```cmdline``` in ```CommandLine``` call a wrapper to that function named ```_update_autonames```. This must be done AFTER mandatory inputs are checked.

Regarding outputs, I set the output values in ```aggregate_outputs```. I'm not 100% sure this is a good call though.

### TODO list

- [ ] Unset files that were not generated
- [x] Multiple replacements in name_template
- [ ] Remove all ```_list_outputs```
- [ ] Implemente name_remove (ie ```nipype.algorithms.misc.Gunzip```)
- [x] Implement name_source with Input/OutputMultiPath (ie ```nipype.algorithms.misc.ModifyAffine```)
- [ ] Implement name_source with several outputs (ie ```nipype.algorithms.misc.Matlab2CSV```)

### List of interfaces that should be updated when this PR works

- [x] nipype.algorithms.misc.TSNR

### CODE REVIEW

File: nipype/interfaces/fsl/utils.py [hunk 1]
```diff
@@ -1579,11 +1579,12 @@ class WarpUtilsInputSpec(FSLCommandInputSpec):
                               desc=('Alternative (to --warpres) specification of the resolution of '
                                     'the output spline-field.'))
 
-    out_file = File(argstr='--out=%s', position=-1, name_source = ['in_file'], output_name='out_file',
-                    desc=('Name of output file. The format of the output depends on what other '
-                          'parameters are set. The default format is a (4D) field-file. If the '
-                          '--outformat is set to spline the format will be a (4D) file of spline '
-                          'coefficients.'))
+    out_file = File(
+        argstr='--out=%s', position=-1, name_source=['in_file'], name_template='%s_coeffs.nii', keep_extension=True,
```
Comment (satra) [inline nipype/interfaces/fsl/utils.py]: why are we forcing the output to be nifti here?

**Comment by chrisfilo** [comment]
please rebase/merge master - we have fixed the CircleCI tests

File: nipype/algorithms/rapidart.py [hunk 2]
```diff
@@ -665,14 +534,122 @@ def _run_interface(self, runtime):
                 nrows.append(mc_in.shape[0])
             matrix = self._get_spm_submatrix(spmmat, sessidx, rows)
             self._stimcorr_core(motparamlist[i], intensityfiles[i],
```
Comment (chrisfilo) [inline nipype/algorithms/rapidart.py]: Codacy Issue found: [Too many positional arguments for method call](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2047345374/issues/source?bid=2767207&fileBranchId=2890884#l536)

Comment (oesteban) [inline nipype/algorithms/rapidart.py]: Codacy only found this?? Then it's useless! :P

Seriously, this is still very preliminary. I am coding with the pylint on, so I also address these issues. I'll let you know when this is ready.

BTW, with this refactoring, we get rid of the infamous ```_list_outputs``` :)

Comment (chrisfilo) [inline nipype/algorithms/rapidart.py]: Codacy Issue found: [Too many positional arguments for method call](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2053828267/issues/source?bid=2767207&fileBranchId=2890884#l536)

File: nipype/interfaces/fsl/epi.py [hunk 3]
```diff
@@ -788,30 +783,29 @@ def _gen_filename(self, name):
...
         else:
-            outputs['vsm_file'] = self._gen_fname(self.inputs.vsm)
+            self.outputs.vsm_file = self._gen_fname(self.inputs.vsm)
         if not isdefined(self.inputs.tmpdir):
             outputs[
                 'exf_mask'] = self._gen_fname(cwd=self._gen_filename('tmpdir'),
                                               basename='maskexf')
         else:
-            outputs['exf_mask'] = self._gen_fname(cwd=self.inputs.tmpdir,
+            self.outputs.exf_mask = self._gen_fname(cwd=self.inputs.tmpdir,
```
Comment (chrisfilo) [inline nipype/interfaces/fsl/epi.py]: Codacy Issue found: [Unexpected keyword argument 'cwd' in method call](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2065943999/issues/source?bid=2767207&fileBranchId=2890884#l806)

File: nipype/interfaces/fsl/epi.py [hunk 4]
```diff
@@ -788,30 +783,29 @@ def _gen_filename(self, name):
...
+                self.outputs.unwarped_file = self._gen_filename('epidw')
         if not isdefined(self.inputs.vsm):
-            outputs['vsm_file'] = self._gen_filename('vsm')
+            self.outputs.vsm_file = self._gen_filename('vsm')
         else:
-            outputs['vsm_file'] = self._gen_fname(self.inputs.vsm)
+            self.outputs.vsm_file = self._gen_fname(self.inputs.vsm)
         if not isdefined(self.inputs.tmpdir):
             outputs[
                 'exf_mask'] = self._gen_fname(cwd=self._gen_filename('tmpdir'),
```
Comment (chrisfilo) [inline nipype/interfaces/fsl/epi.py]: Codacy Issue found: [Unexpected keyword argument 'cwd' in method call](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2065943999/issues/source?bid=2767207&fileBranchId=2890884#l803)

File: nipype/interfaces/fsl/epi.py [hunk 5]
```diff
@@ -264,15 +263,14 @@ def _list_outputs(self):
                 base_path = None
         else:
             base = split_filename(self.inputs.in_file)[1] + '_base'
-        outputs['out_fieldcoef'] = self._gen_fname(base, suffix='_fieldcoef',
+        self.outputs.out_fieldcoef = self._gen_fname(base, suffix='_fieldcoef',
                                                    cwd=base_path)
-        outputs['out_movpar'] = self._gen_fname(base, suffix='_movpar',
+        self.outputs.out_movpar = self._gen_fname(base, suffix='_movpar',
```
Comment (chrisfilo) [inline nipype/interfaces/fsl/epi.py]: Codacy Issue found: [Unexpected keyword argument 'ext' in method call](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2065943999/issues/source?bid=2767207&fileBranchId=2890884#l268)

File: nipype/interfaces/fsl/epi.py [hunk 6]
```diff
@@ -264,15 +263,14 @@ def _list_outputs(self):
                 base_path = None
         else:
             base = split_filename(self.inputs.in_file)[1] + '_base'
-        outputs['out_fieldcoef'] = self._gen_fname(base, suffix='_fieldcoef',
+        self.outputs.out_fieldcoef = self._gen_fname(base, suffix='_fieldcoef',
```
Comment (chrisfilo) [inline nipype/interfaces/fsl/epi.py]: Codacy Issue found: [Unexpected keyword argument 'cwd' in method call](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2065943999/issues/source?bid=2767207&fileBranchId=2890884#l266)

File: nipype/interfaces/fsl/model.py [hunk 7]
```diff
@@ -693,24 +690,23 @@ def _get_numcons(self):
...
+        self.outputs.residual4d = self._gen_fname('res4d.nii', cwd=results_dir)
+        self.outputs.dof_file = os.path.join(results_dir, 'dof')
+        self.outputs.sigmasquareds = self._gen_fname('sigmasquareds.nii',
                                                    cwd=results_dir)
-        outputs['thresholdac'] = self._gen_fname('threshac1.nii',
+        self.outputs.thresholdac = self._gen_fname('threshac1.nii',
                                                  cwd=results_dir)
         if Info.version() and LooseVersion(Info.version()) < LooseVersion('5.0.7'):
-            outputs['corrections'] = self._gen_fname('corrections.nii',
+            self.outputs.corrections = self._gen_fname('corrections.nii',
```
Comment (chrisfilo) [inline nipype/interfaces/fsl/model.py]: Codacy Issue found: [Unexpected keyword argument 'cwd' in method call](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2065944001/issues/source?bid=2767207&fileBranchId=2890884#l707)

File: nipype/algorithms/rapidart.py (hunk 2 above)
Comment (chrisfilo) [inline nipype/algorithms/rapidart.py]: Codacy Issue found: [Too many positional arguments for method call](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/2072986286/issues/source?bid=2767207&fileBranchId=2890884#l536)

**Pull request closed without merging**